File: ruby/lang/security/md5-used-as-password.rb
require 'digest'

#### True Positives ####
def ex1 (user, pwtext)
    # ruleid: md5-used-as-password
    user.set_password Digest::MD5.hexdigest pwtext
end

def ex2 (user, pwtext)
    md5 = Digest::MD5.new
    md5.update pwtext
    md5 << salt(pwtext)
    dig = md5.hexdigest
    # ruleid: md5-used-as-password
    user.set_password dig
end

#### True Negatives ####
def ok1 (user, pwtext)
    # ok: md5-used-as-password
    user.set_password Digest::SHA256.hexdigest pwtext
end

def ok2 (user, pwtext)
    sha = Digest::SHA256.new
    sha.update pwtext
    sha << salt(pwtext)
    dig = sha.hexdigest
    # ok: md5-used-as-password
    user.set_password dig
end

File: php/lang/security/md5-used-as-password.php
<?php

function test1($value) {
    $pass = md5($value);
    // ruleid: md5-used-as-password
    $user->setPassword($pass);
}

function test2($value) {
    $pass = hash('md5', $value);
    // ruleid: md5-used-as-password
    $user->setPassword($pass);
}

function okTest1($value) {
    // ok: md5-used-as-password
    $pass = hash('sha256', $value);
    $user->setPassword($pass);
}
